Create example scenario for usage of GitHub Connector (#52)
kyma-incubator/hack-showcase: docs/examples/demoscenario.md

	- [Option 1. Through Kyma user interface](#option-1-through-kyma-user-interface)
		- [Binding application to the namespace](#binding-application-to-the-namespace)
		- [Setting up ServiceInstance and Lambda](#setting-up-serviceinstance-and-lambda)
		- [Creation of Lambda function](#creation-of-lambda-function)
	- [Option 2. Shell script](#option-2-shell-script)

## Overview

This purpose of this guide is to show example usage of GitHub Connector, which allows you to to handle GitHub events. With it's help you will create a lamda in Kyma, which reacts to new issue on connected repository with a short message accessible in lambda's logs. 

## Prerequisites

- Kyma with GitHub Connector deployed on your cluster (see [installation guide](helm-installation-tutorial.md)).
- WebHook configured to deliver payload to the Connector's `/webhook` endpoint.

## Configuration of ServiceInstance to use Lambda functions

To write lambda in Kyma that can utilise the functionalities of GitHub Connector you must configure the application binding and ServiceInstance.

### Option 1. Through Kyma user interface

> #### Binding application to the namespace

Reviewer: It's a convention in our docs - "Namespace" is a capitalized term project-wide.

2. Run the script and supply the name, which has to be the same as the one you used as release name while installing GitHub Connector with helm, and the namespace in which GitHub Connector application is running.

   ```shell
   sh demoscenario.sh {NAME} {NAMESPACE}
   ```

   After you trigger the script you get the following output that shows the created resources:

   ```
   applicationmapping.applicationconnector.kyma-project.io/gh-connector-example-app created
   serviceinstance.servicecatalog.k8s.io/gh-connector-example created
   function.kubeless.io/gh-connector-example-lambda created
   subscription.eventing.kyma-project.io/gh-connector-example-lambda-issuesevent-opened-v1 created
   Subscribed! Happy GitHub Connecting!
   ```

   Now your GitHub Connector is configured to react to new issues opened on GitHub repository you have connected to during Connector installation.

   To test if connetion works, create a new issue on connected repository to trigger the event in lambda. Then, run this command to find lambda's name (it should consist of name you have provided earlier and *-lambda* suffix):

   `kubectl get pods -n {NAMESPACE}`

>    Copy the output and paste it into this command to check lamda's logs:

Reviewer: The output of `kubectl get pods -n {NAMESPACE}` is the list of all Pods in the Namespace. Are you sure I should copy this entire list over to the next command?

Author: Good point, fixed.

# Using GitHub Connector <!-- omit in toc -->

- [Overview](#overview)
- [Prerequisites](#prerequisites)
- [Steps](#steps)

## Overview

This purpose of this guide is to show example usage of GitHub Connector, which allows you to to handle GitHub events. With its help, you will create a lamda in Kyma, which reacts to new issues created in the connected repository with a short message accessible in lambda's logs. 

## Prerequisites

> - Kyma with GitHub Connector deployed on your cluster (see [installation guide](/chart/githubconnector/README.md)).

Reviewer: ```suggestion
- Kyma with the GitHub Connector [installed](/chart/githubconnector/README.md)).
```

> - WebHook configured to deliver payload to the Connector's `/webhook` endpoint.

Reviewer: How do I do that?

Author: I assumed that user has certain level of basic knowledge, but for the peace of mind I'll add a link to GitHub's website where it's explained.

### Steps

1. Go to the `/chart/demoscenario` directory, where you can find `demoscenario.sh` script.

> 2. Run the script and supply the name, which has to be the same as the one you used as release name while installing GitHub Connector with helm, and the Namespace in which GitHub Connector application is running.

Reviewer: ```suggestion
2. Run the script that sets the demo scenario up. Pass the name of the Helm release of the GitHub Connector and the Namespace in which you installed it. Run:
```
Do you just pass the Namespace/release name without any flags? Just like: 
`sh demoscenario.sh githubconnector default` ?

Author: Yes, for now. It will be changed in the future.

   ```shell
   sh demoscenario.sh {NAME} {NAMESPACE}
   ```

   After you trigger the script you get the following output that shows the created resources:

   ```
   applicationmapping.applicationconnector.kyma-project.io/gh-connector-example-app created
   serviceinstance.servicecatalog.k8s.io/gh-connector-example created
   function.kubeless.io/gh-connector-example-lambda created
   subscription.eventing.kyma-project.io/gh-connector-example-lambda-issuesevent-opened-v1 created
   Subscribed! Happy GitHub Connecting!
   ```

   Now your GitHub Connector is configured to react to new issues opened on GitHub repository you have connected to during Connector installation.

> 3. To test if connetion works, create a new issue on connected repository to trigger the event in lambda. Then, run this command to find name of the Pod in which lambda is running:

Reviewer: I guess this could be split into two separate steps. 
1. Create an issue.
2. "Get the name of the Pod running the lamda function. Run this command: `kubectl get pods --all-namespaces | grep "lambda"`. 

If you get Pods from all Namespaces and search for "lambda" you won't have to pass the Namespace in the call. It's also more convenient as you don't have to look for the Pod among the names of other Pods.

Author: I'll split it into two steps and add grep command, but I don't think omitting a Namespace is a good idea - if user have more lambda's in other Namespaces it could be difficult for him to find the proper one.

Reviewer: good point

# Using GitHub Connector <!-- omit in toc -->

- [Overview](#overview)
- [Prerequisites](#prerequisites)
- [Steps](#steps)

## Overview

> This purpose of this guide is to show example usage of GitHub Connector, which allows you to to handle GitHub events. With it's help you will create a lamda in Kyma, which reacts to new issue on connected repository with a short message accessible in lambda's logs. 

Reviewer: ```suggestion
This purpose of this guide is to show example usage of GitHub Connector, which allows you to to handle GitHub events. With its help, you will create a lamda in Kyma, which reacts to new issues created in the connected repository with a short message accessible in lambda's logs. 
```

it's = it is
its = belonging or pertaining to that thing

The sentence that starts with `with its...` is supposed to describe what the user does and achieves in this example, right? I think that it's all about connecting Kyma to your Github repo, then creating a lambda and testing the connection.

> 

Reviewer: Redundant newline - one's enough ;)
Same applies to other newlines between headings.

## Prerequisites

- Kyma with GitHub Connector deployed on your cluster (see [installation guide](/chart/githubconnector/README.md).
- WebHook configured to deliver payload to the Connector's `/webhook` endpoint.

### Steps

1. Go to the `/chart/demoscenario` directory, where you can find `demoscenario.sh` script.

2. Run the script and supply the name, which has to be the same as the one you used as release name while installing GitHub Connector with helm, and the Namespace in which GitHub Connector application is running.

   ```shell
   sh demoscenario.sh {NAME} {NAMESPACE}
   ```

   After you trigger the script you get the following output that shows the created resources:

   ```
   applicationmapping.applicationconnector.kyma-project.io/gh-connector-example-app created
   serviceinstance.servicecatalog.k8s.io/gh-connector-example created
   function.kubeless.io/gh-connector-example-lambda created
   subscription.eventing.kyma-project.io/gh-connector-example-lambda-issuesevent-opened-v1 created
   Subscribed! Happy GitHub Connecting!
   ```

>    Now your GitHub Connector is configured to react to new issues opened on GitHub repository you have connected to during Connector installation.

Reviewer: Follow the "steps" convention and make sure all steps are entries in the ordered list you started with steps 1 and 2.

> 3. To test if connetion works, create a new issue on connected repository to trigger the event in lambda. Then, run this command to find lambda's name:

Reviewer: You're running this command to find the name of the Pod that contains the lambda function, right? Make sure it's clear in this sentence.

Reviewer 2: You're right. Corrected that.

   `kubectl get pods -n {NAMESPACE}`

> 4. Copy the lambda's name from the output (it should consist of name you have provided earlier and `-lambda` suffix) and paste it into this command to check it's logs:

Reviewer: You don't have to say that the user has to copy the name of the lambda's Pod. Maybe this is used by an old-school geek who'd rather input everything manually than copy things over? 

This could be much shorter. 
> Run this command to get the logs from the Pod that runs the lambda function and search for the "issue opened" phrase to verify if (...) 
`kubectl logs -n {NAMESPACE} {LAMBDA-NAME} | grep "issue opened"`

   `kubectl logs -n {NAMESPACE} {LAMBDA-NAME}`

>    After you run the command you should find a line containg phrase `Issue opened`. This means your lambda reacted to new issue opened event and you configured everything properly. 

Reviewer: If you followed my previous suggestion then you don't really need this sentence anymore :)

# Usage of GitHub Connector <!-- omit in toc -->

- [Prerequisites](#prerequisites)
- [Configuration of ServiceInstance to use Lambda functions](#configuration-of-serviceinstance-to-use-lambda-functions)
	- [Option 1. Through Kyma user interface](#option-1-through-kyma-user-interface)
		- [Binding application to the namespace](#binding-application-to-the-namespace)
		- [Setting up ServiceInstance and Lambda](#setting-up-serviceinstance-and-lambda)
		- [Creation of Lambda function](#creation-of-lambda-function)
	- [Option 2. Shell script](#option-2-shell-script)

## Prerequisites

- Kyma with GitHub Connector deployed on your cluster (see [installation guide](helm-installation-tutorial.md)).

> - WebHook configured to deliver payload to the Connector's ```/webhook``` endpoint.

Reviewer 2: Thanks!

## Configuration of ServiceInstance to use Lambda functions

To write lambda in Kyma that can utilise the functionalities of GitHub Connector you must configure the application binding and ServiceInstance.

> ### Option 1. Through Kyma user interface

Reviewer: I know that you must've spent quite some time describing this flow and capturing screenshots but I suggest that you get rid of this section altogether. In Kyma docs the CLI is a first-class citizen, meaning that we document getting from point a to point b only through the terminal. There are exceptions of course as a few components and aspects of Kyma can be adjusted only through the UI. 

Furthermore, why would I want to use this flow instead of simply running a single script you guys prepared? ;)

Reviewer 2: We decided to keep that because: firstly - one may have troubles running this script on Windows, secondly - the graphics tutorial shows user what is actually created during configuration process and they can adjust the process to their own needs.

Reviewer: https://kyma-project.io/community/guidelines/content/#content-strategy-content-strategy-the-main-purpose-of-instructions

> The main purpose of instructions
One of the main Kyma principles is that CLI is, metaphorically speaking, the first-level citizen. Therefore, the documentation's main focus is to explain concepts and provide step-by-step instructions using CLI commands instead of Console UI screenshots. The only exception from the rule applies to the components that cannot be managed through the CLI, in which case screenshots are essential to explain specific functionalities.

Reviewer: Your screenshots don't follow the guidelines either: 

https://kyma-project.io/community/guidelines/content/#screenshots-screenshots

Reviewer: If you want to explain what your script does you can describe it here - this approach has been done in our docs on numerous occasions. 
As for customization options - you can use comments in your script. Describe the tweakable parameters in the doc, link to specific lines in the .sh file.

Reviewer: The fact that I have to jump between your doc and the UI to see what actions should I perform is a nuisance and is likely to scare people away rather than attract them to trying your connector.

Reviewer 2: We removed this guide.

#### Binding application to the namespace

1. Choose namespace created during installation of GitHub Connector.
	![Choose namespace](./pictures/demoscenario-01-choose-namespace.png)

> 2. Click **```Show All Applications```**.

Reviewer 2: Custom solution - I think it looks like closer to the button user will click.

Anyway - ok.

> 2. Click **```Show All Applications```**.

Reviewer 2: I'll keep it in mind.

#### Creation of Lambda function

After clicking **```+ Add Lambda```** you will be able to select triggering options and write body of your function.

1. Choose function name and click **```+ Select Funtion Trigger```** to choose the trigger method.
	![Select function trigger](./pictures/demoscenario-12-select-function-trigger.png)
2. From the drop-down menu that appears choose the trigger option you desire. For the purpose of this example **Event Trigger** is chosen.
	![Choose function trigger](./pictures/demoscenario-13-choose-function-trigger.png)
3. Select events for which this specific lambda you desire to be able to react.
	![Select events](./pictures/demoscenario-14-choose-events.png)
4. Create service binding and choose your service. Click **```Create Service Binding```** to apply binding.
   	![Write lambda](./pictures/demoscenario-15-create-service-binding.png)

5. Write body of your function and click **```Create```**.
	![Write lambda](./pictures/demoscenario-16-write-lambda.png)

Everything should be working now.

### Option 2. Shell script

1. Go to the `/chart/demoscenario` directory, where you can find `demoscenario.sh` script.

> 2. Run the script and supply the name, which has to be the same as the one you used as release name while installing GitHub Connector with helm, and the namespace in which GitHub Connector application is running (github-connector).

Reviewer: Rephrase this sentence having the comments below in mind. See kyma-project.io docs for inspiration.

   ```shell

>    sh demoscenario.sh gh-connector-example github-connector

Reviewer: This command should have placeholders to show the potential user which parameter should be put in which position. We introduce placeholders in curly braces `{}`. The release name of your connector is something that won't be changed I guess (why would anyone want to do it?), so I'd make the namespace a variable.

Reviewer 2: Actually there is a possibility to change the name of release. You may want to have multiple connectors deployed :P

>    sh demoscenario.sh gh-connector-example github-connector

Reviewer: ```suggestion
   sh demoscenario.sh gh-connector-example {NAMESPACE}
```

   ```

>    You can see output similar to the following:

Reviewer: ```suggestion
   After you trigger the script you get the following output that shows the created resources:
```

   ```
   applicationmapping.applicationconnector.kyma-project.io/gh-connector-example-app created
   serviceinstance.servicecatalog.k8s.io/gh-connector-example created
   function.kubeless.io/gh-connector-example-lambda created
   subscription.eventing.kyma-project.io/gh-connector-example-lambda-issuesevent-opened-v1 created
   Subscribed! Happy GitHub Connecting!
   ```

>    Lambda reacts to new issues opened in GitHub repository you are bound to. It logs `Issue opened` when someone creates a new issue.

Reviewer: After I don't really get what you mean here. As I commented previously, your documentation lacks an explanation of what the connector does for the user. You could include a short summary of the expected outcome in the introductory paragraph of this scenario.

Reviewer: I'm missing one more step here. You should describe the steps the user needs to take to test if this connection actually works. Create an issue and then do what? Get logs I guess, but where from? I think it's important to include such information to make this a comprehensive demo.

Reviewer 2: We added it.
